src/tls/ffi.c

```diff
@@ -0,0 +1,402 @@
+        close(fdout);
+        free(out_name);
+        free(err_name);
+        return;
+    }
+    r->fd_out_new = fdout;
+    r->fd_err_new = fderr;
+    r->out_name = out_name;
+    r->err_name = err_name;
+    
+    // Capture current stdout/stderr
+    r->fd_out_old = dup(STDOUT_FILENO);
+    r->fd_err_old = dup(STDERR_FILENO);
+    
+    // Redirect stdout and stderr to the temp files
+    dup2(fdout, STDOUT_FILENO);
+    dup2(fderr, STDERR_FILENO);
+}
+
+#define C_ASSERT(e) typedef char __C_ASSERT__[(e)?1:-1]
+C_ASSERT(sizeof(size_t) == sizeof(value));
+
+// Called by the host app to configure miTLS ahead of creating a connection
+int FFI_mitls_configure(mitls_state **state, const char *tls_version, const char *host_name, char **outmsg, char **errmsg)
+{
+    int ret = 0;
+    CAMLlocal3(config, version, host);
+    mitls_redirect r;
+
+    *state = NULL;
+    *outmsg = NULL;
+    *errmsg = NULL;
+    redirect_stdio(&r);
+    
+    version = caml_copy_string(tls_version);  
+    host = caml_copy_string(host_name);
+    config = caml_callback2_exn(*g_mitls_FFI_Config, version, host);
+    if (Is_exception_result(config)) {
+        printf("Exception!  %s\n", caml_format_exception(Extract_exception(config)));
+    } else {
+        mitls_state * s;
+        
+        // Allocate space on the heap, to store an OCaml value
+        s = (mitls_state*)malloc(sizeof(mitls_state));
+        if (s) {
+            // Tell the OCaml GC about the heap address, so it is treated
+            // as a GC root, keeping the config object live.
+            s->fstar_state = config; 
+            caml_register_generational_global_root(&s->fstar_state);
+            if (strcmp(tls_version, "1.3") == 0) {
+                s->connect_callback = g_mitls_FFI_Connect13;
+            } else {
+                s->connect_callback = g_mitls_FFI_Connect12;
+            }
+            *state = s;
+            ret = 1;
+        }
+    }
+    
+    restore_stdio(&r, outmsg, errmsg);
+    return ret;
+}
+
+// Called by the host app to free a mitls_state allocated by FFI_mitls_configure()
+void FFI_mitls_close(mitls_state *state)
+{
+    if (state) {
+        caml_remove_generational_global_root(&state->fstar_state);
+        state->fstar_state = 0;
+        free(state);
+    }
+}
+
+void FFI_mitls_free_msg(char *msg)
+{
+    free(msg);
+}
+
+void FFI_mitls_free_packet(void *packet)
+{
+    free(packet);
+}
+
+void * copypacket(value packet, /* out */ size_t *packet_size)
+{
+    void *p;
+    mlsize_t size;
+        
+    size = caml_string_length(packet);
+    p = malloc(size);
+    if (p) {
+        memcpy(p, String_val(packet), size);
+        *packet_size = size;
+    }
+    return p;
+}
+
+// Called from FStar code to send via TCP
+CAMLprim value ocaml_send_tcp(value cookie, value bytes)
+{
+    mlsize_t buffer_size;
+    char *buffer;
+    int retval;
+    size_t c = Long_val(cookie);
+    struct _FFI_mitls_callbacks *callbacks = (struct _FFI_mitls_callbacks *)c;
```

Comment:
If my understanding is correct, you're storing a pointer within an OCaml `int` value. I'm actually amazed that this works! This means the pointer always fits within 63 bits (or 31 bits for 32-bit builds of OCaml)...